wire-ios-sync-engine/Source/Repositories/TeamRepository.swift

```diff
@@ -0,0 +1,261 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireAPI
+import WireDataModel
+
+/// Facilitate access to team related domain objects.
+///
+/// A repository provides an abstraction for the access and storage
+/// of domain models, concealing how and where the models are stored
+/// as well as the possible source(s) of the models.
+
+protocol TeamRepositoryProtocol {
+
+    /// Pull self team metadata frmo the server and store locally.
+
+    func pullSelfTeam() async throws
+
+    /// Pull team roles for the self team from the server and store locally.
+
+    func pullSelfTeamRoles() async throws
+
+    /// Pull team members for the self team from the server and store locally.
+
+    func pullSelfTeamMembers() async throws
+
+    /// Fetch the legalhold status for the self user from the server.
+
+    func fetchSelfLegalholdStatus() async throws -> LegalholdStatus
+
+}
+
+final class TeamRepository: TeamRepositoryProtocol {
+
+    private let selfTeamID: UUID
+    private let teamsAPI: any TeamsAPI
+    private let context: NSManagedObjectContext
+
+    init(
+        selfTeamID: UUID,
+        teamsAPI: any TeamsAPI,
+        context: NSManagedObjectContext
+    ) {
+        self.selfTeamID = selfTeamID
+        self.teamsAPI = teamsAPI
+        self.context = context
+    }
+
+    // MARK: - Pull self team
+
+    func pullSelfTeam() async throws {
+        let team = try await fetchSelfTeamRemotely()
+        await storeTeamLocally(team)
+    }
+
+    private func fetchSelfTeamRemotely () async throws -> WireAPI.Team {
+        do {
+            return try await teamsAPI.getTeam(for: selfTeamID)
+        } catch {
+            throw TeamRepositoryError.failedToFetchRemotely(error)
+        }
+    }
+
+    private func storeTeamLocally(_ teamAPIModel: WireAPI.Team) async {
+        await context.perform { [context] in
+            let team = WireDataModel.Team.fetchOrCreate(
+                with: teamAPIModel.id,
+                in: context
+            )
+
+            let selfUser = ZMUser.selfUser(in: context)
```

Comment:
Using this func is kind of a smelly piece of code, because it has many implicit conditions. Can we avoid it somehow in new code?
Maybe one idea could be to write a `SelfUserRepository` and inject it here to make it easier to refactor in future?

Reply:
Certainly!

Reply:
Hmmm so we already have a `UserRepository` define in WireDataModel (because it's used in WireRequestStrategy). The way it's used now is problematic so I'll try to fix that in a separate PR first so we can use it here.

Reply:
done in 3f9ef3da990450c0801e62bcc32fa94459d1ef25